File: packages/protocol-types/src/actions.ts
/**
 * Action types — what a player can do at each state.
 *
 * Per `spec/state-machine.md` §3 and `spec/tx-types.md` §1.
 */

import type {
  ActionNonce,
  GameId,
  Hash256,
  Pubkey33,
  RoundNumber,
  Satoshis,
  Seat,
} from './primitives.js';
import type { ConcealedCard, RevealProof } from './cards.js';

/** Every action type a player or auto-trigger can produce. */
export type ActionType =
  | 'Join'
  | 'EntropyCommit'
  | 'EntropyReveal'
  | 'CardReveal'
  | 'DealConcealed'
  | 'BetAction'
  | 'Pass'
  | 'Fold'
  | 'Settle'
  | 'RotateTurn'
  | 'TableLock'
  | 'TableClose'
  | 'Timeout'
  | 'Recovery';

/**
 * Base shape every signed action commits to. The signature scope per
 * `spec/serialisation.md` §6 covers the canonical encoding of this
 * record with the signature field zeroed.
 */
export interface SignedActionBase {
  readonly game_id: GameId;
  readonly round_number: RoundNumber;
  readonly referenced_state_hash: Hash256;
  readonly action_type: ActionType;
  readonly action_nonce: ActionNonce;
  readonly acting_player_seat: Seat | null;
  readonly authorising_signature: string;
  /** Hash of the state the action transitions to. */
  readonly successor_state_commitment: Hash256;
}

/** Join action carries the player's pubkey and stake amount. */
export interface JoinAction extends SignedActionBase {
  readonly action_type: 'Join';
  readonly player_pubkey: Pubkey33;
  readonly stake_amount: Satoshis;
}

/** Entropy commitment publishes only the hash; reveal comes later. */
export interface EntropyCommitAction extends SignedActionBase {
  readonly action_type: 'EntropyCommit';
  readonly commitment_hash: Hash256;
}

/** Entropy reveal opens the prior commitment. */
export interface EntropyRevealAction extends SignedActionBase {
  readonly action_type: 'EntropyReveal';
  readonly entropy: Hash256;
}

/** Card reveal opens one deck position. */
export interface CardRevealAction extends SignedActionBase {
  readonly action_type: 'CardReveal';
  readonly reveal: RevealProof;
}

/**
 * DealConcealed populates `RoundState.concealed_deck` at the
 * S5 DECK_COMMITTED transition. The orchestrator supplies one
 * `ConcealedCard` per deck position; the engine validates the
 * count and uniqueness of positions, then stamps the deck onto
 * the round state. Subsequent `Fold` actions then have a deck
 * to act on.
 *
 * Legal at S5_DECK_COMMITTED only; rejected if a concealed deck
 * is already populated (idempotency guard).
 */
export interface DealConcealedAction extends SignedActionBase {
  readonly action_type: 'DealConcealed';
  readonly concealed_cards: readonly ConcealedCard[];
}

/** Bet action commits a bet amount at S8 BET_DECISION. */
export interface BetAction extends SignedActionBase {
  readonly action_type: 'BetAction';
  readonly bet_amount: Satoshis;
}

/** Pass — no economic change, used as legitimate default at S8. */
export interface PassAction extends SignedActionBase {
  readonly action_type: 'Pass';
}

/** Fold — extended model only; not used by In-Between v1. */
export interface FoldAction extends SignedActionBase {
  readonly action_type: 'Fold';
}

/** Settle — resolve the round outcome. */
export interface SettleAction extends SignedActionBase {
  readonly action_type: 'Settle';
}

/** Rotate to next player. */
export interface RotateTurnAction extends SignedActionBase {
  readonly action_type: 'RotateTurn';
}

/** Lock the table once player_count_min met. */
export interface TableLockAction extends SignedActionBase {
  readonly action_type: 'TableLock';
}

/** Close the table after the final round. */
export interface TableCloseAction extends SignedActionBase {
  readonly action_type: 'TableClose';
}

/** Timeout default — broadcast by any participant past the decision deadline. */
export interface TimeoutAction extends SignedActionBase {
  readonly action_type: 'Timeout';
  readonly default_consequence: ActionType;
  readonly silenced_seat: Seat;
}

/** Recovery — pre-signed tx whose `nLockTime` is the absolute recovery height. */
export interface RecoveryAction extends SignedActionBase {
  readonly action_type: 'Recovery';
  readonly recovery_trigger: string;
}

/** Discriminated union over every action type. */
export type SignedAction =
  | JoinAction
  | EntropyCommitAction
  | EntropyRevealAction
  | CardRevealAction
  | DealConcealedAction
  | BetAction
  | PassAction
  | FoldAction
  | SettleAction
  | RotateTurnAction
  | TableLockAction
  | TableCloseAction
  | TimeoutAction
  | RecoveryAction;

File: apps/relay-go/pkg/types/actions.go
package types

// ActionType is the discriminant tag of every protocol action.
type ActionType string

const (
	ActionJoin           ActionType = "Join"
	ActionEntropyCommit  ActionType = "EntropyCommit"
	ActionEntropyReveal  ActionType = "EntropyReveal"
	ActionCardReveal     ActionType = "CardReveal"
	ActionBet            ActionType = "BetAction"
	ActionPass           ActionType = "Pass"
	ActionFold           ActionType = "Fold"
	ActionSettle         ActionType = "Settle"
	ActionRotateTurn     ActionType = "RotateTurn"
	ActionTableLock      ActionType = "TableLock"
	ActionTableClose     ActionType = "TableClose"
	ActionTimeout        ActionType = "Timeout"
	ActionRecovery       ActionType = "Recovery"
)

// SignedAction is a single Go struct carrying every action variant's
// fields. Optional fields are zero-valued for variants that do not
// use them. We deliberately collapse the discriminated union here
// because Go's encoding/json does not natively support sum types and
// the unified shape is easier for the engine to dispatch on
// ActionType.
type SignedAction struct {
	GameId                    GameId      `json:"game_id"`
	RoundNumber               RoundNumber `json:"round_number"`
	ReferencedStateHash       Hash256     `json:"referenced_state_hash"`
	ActionType                ActionType  `json:"action_type"`
	ActionNonce               ActionNonce `json:"action_nonce"`
	ActingPlayerSeat          *Seat       `json:"acting_player_seat"` // pointer so JSON null is preserved
	AuthorisingSignature      string      `json:"authorising_signature"`
	SuccessorStateCommitment  Hash256     `json:"successor_state_commitment"`

	// Variant-specific fields. Zero-valued when not applicable.
	PlayerPubkey       Pubkey33    `json:"player_pubkey,omitempty"`
	StakeAmount        Satoshis    `json:"stake_amount,omitempty"`
	CommitmentHash     Hash256     `json:"commitment_hash,omitempty"`
	Entropy            Hash256     `json:"entropy,omitempty"`
	Reveal             RevealProof `json:"reveal"`
	BetAmount          Satoshis    `json:"bet_amount,omitempty"`
	DefaultConsequence ActionType  `json:"default_consequence,omitempty"`
	SilencedSeat       *Seat       `json:"silenced_seat,omitempty"`
	RecoveryTrigger    string      `json:"recovery_trigger,omitempty"`
}
